metadata/snapshot.go

```diff
@@ -274,7 +275,22 @@ func (s *snapshotter) Mounts(ctx context.Context, key string) ([]mount.Mount, er
 }
 
 func (s *snapshotter) Prepare(ctx context.Context, key, parent string, opts ...snapshots.Opt) ([]mount.Mount, error) {
-	return s.createSnapshot(ctx, key, parent, false, opts)
+	mounts, err := s.createSnapshot(ctx, key, parent, false, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if s.db.dbopts.publisher != nil {
+		if err := s.db.dbopts.publisher.Publish(ctx, "/snapshot/prepare", &eventstypes.SnapshotPrepare{
+			Key:         key,
+			Parent:      parent,
+			Snapshotter: s.name,
+		}); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
How about having `s.publish(ctx, "/snapshot/prepare", ...)`? The method could early-return if s.db.dbopts.publisher is nil.

Reply:
It doesn't make sense to do that here since `s.db.dbopts.publisher == nil` would end up unnecessarily allocating the event type. If they were consolidated to a single function, then it would end up with a giant switch statement with more total lines of code. Since they are each used exactly once in this context, inline makes sense.